cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,1164 @@
+
+    // FIXME: Perform nbr_intersection in chuncks.
+    auto [intersection_offsets, intersection_indices] =
+      detail::nbr_intersection(handle,
+                               cur_graph_view,
+                               cugraph::edge_dummy_property_t{}.view(),
+                               edges,
+                               edges + num_edges,
+                               std::array<bool, 2>{true, true},
+                               do_expensive_check);
+  
+    rmm::device_uvector<vertex_t> num_triangles(num_edges, handle.get_stream());
+    thrust::fill(
+      handle.get_thrust_policy(), num_triangles.begin(), num_triangles.end(), size_t{0});
+
+    // Update the number of triangles of each (p, q) edges by looking at their intersection
+    // size
+    thrust::adjacent_difference(handle.get_thrust_policy(),
+                                intersection_offsets.begin() + 1,
+                                intersection_offsets.end(),
+                                num_triangles.begin());
+
+    thrust::for_each(handle.get_thrust_policy(),
+                     thrust::make_counting_iterator<edge_t>(0),
+                     thrust::make_counting_iterator<edge_t>(intersection_indices.size()),
+                     extract_p_r<vertex_t, edge_t, decltype(edges)>{
+                     num_edges,
+                     raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+                     raft::device_span<vertex_t const>(intersection_indices.data(), intersection_indices.size()),
+                     raft::device_span<vertex_t>(num_triangles.data(), num_triangles.size()),
+                     edges});
+    
+    thrust::for_each(handle.get_thrust_policy(),
+                     thrust::make_counting_iterator<edge_t>(0),
+                     thrust::make_counting_iterator<edge_t>(intersection_indices.size()),
+                     extract_q_r<vertex_t, edge_t, decltype(edges)>{
+                     num_edges,
+                     raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+                     raft::device_span<vertex_t const>(intersection_indices.data(), intersection_indices.size()),
+                     raft::device_span<vertex_t>(num_triangles.data(), num_triangles.size()),
+                     edges});
+
+    auto edges_to_num_triangles = thrust::make_zip_iterator(
+      edges, num_triangles.begin());
+
+    // 'invalid_edge_first' marks the beginning of the edges to be removed
+    auto invalid_edge_last =
+      thrust::stable_partition(handle.get_thrust_policy(),
+                               edges_to_num_triangles,
+                               edges_to_num_triangles + num_triangles.size(),
+                               [k] __device__(auto e) {
+                                 auto num_triangles = thrust::get<1>(e);
+                                 return num_triangles < k - 2;
+                               });
```

Comment:
> If we don't need to maintain the sorted order here

We actually need the sorted order because I am keeping `edges_to_num_triangles` sorted so that I can run a binary search on the original edge list later. I have a separate array keeping track of the edge list sorted by destination.